### ISSUE TITLE
ssl-default-bind-ciphersuites influencing TLS1.2.

### ISSUE THREAD

**Issue by andreas-p** [haproxy/haproxy#2459]
### Detailed Description of the Problem

As per Documentation at https://www.haproxy.com/documentation/haproxy-configuration-manual/latest/#3.1-ssl-default-bind-ciphersuites, ssl-default-bind-ciphersuites should affect TLS1.3 only, but it also affects TLS1.2.

### Expected Behavior

Documentation consistent with actual behaviour

### Steps to Reproduce the Behavior

I noticed this while upgrading from 2.4 to 2.6, when an ancient client stopped working. Examining the packets using Wireshark, I found that the client would connect using TLS1.2, the server will negotiate TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 with 2.4 (allowed by `ssl-default-bind-ciphers DHE-RSA-AES128-GCM-SHA256`), but not with 2.6 or 2.8.
Adding `ssl-default-bind-ciphersuites (..):TLS_RSA_WITH_AES_256_GCM_SHA384`  made the client work again, using that ciphersuite. Per documentation, this shouldn't work. (nb: `add-default-bind-cipher (..):AES256-GCM-SHA384` works as well, which is conforming to the documentation).

This was checked on Debian12 with self-compiled versions 2.4.25 and 2.6.16, using the identical build options and debian openssl 3.0.11, and haproxy_2.8.5-1~bpo12+1_amd64.deb
```


### Output of `haproxy -vv`

```plain
HAProxy version 2.8.5-1~bpo12+1 2023/12/09 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.5.html
Running on: Linux 6.1.0-17-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.69-1 (2023-12-30) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_OT=1 USE_QUIC=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_QUIC_OPENSSL_COMPAT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL +OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION +QUIC +QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 3.0.11 19 Sep 2023
Running on OpenSSL version : OpenSSL 3.0.11 19 Sep 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with OpenTracing support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[  OT] opentracing
	[SPOE] spoe
	[TRACE] tracekjoj
```

**Comment by wlallemand** [comment]
Unfortunately we can't do much about it, haproxy uses the ciphersuites API of OpenSSL which is documented to works for TLSv1.3 ciphersuites https://www.openssl.org/docs/man3.2/man3/SSL_CTX_set_ciphersuites.html. Since TLS_RSA_WITH_AES_256_GCM_SHA384 is a TLSv1.2 cipher it's possible that it was simply converted by OpenSSL to the cipher list.

**Comment by andreas-p** [comment]
Reading that openssl docs, they don't really say what's happening with set_ciphersuites for TLS1.2. 
I guess it's just the other way: the cipher_list is converted to a ciphersuite. Ultimately, that's what the TLS protocol agrees about: the ciphersuite, not the individual ciphers. From observing the protocol handshake, I naively set the TLS1.2 ciphersuite and voila it worked.
So the docs should probably read: cipher_list is <=TLS1.2 only, ciphersuite for all TLS (at least >=1.2, no more TLS1 around hopefully...)

**Comment by wlallemand** [comment]
Well, as usual the OpenSSL documentation is unclear :-)  The SSL_CTX_set_ciphersuites() function appeared with OpenSSL 1.1.1 and TLSv1.3. it didn't exist with OpenSSL 1.0.2 so this format was not supported until TLS1.3 appeared in OpenSSL. I'll change the doc to be more generic about ciphersuites as they are described in the TLS RFC.

**Comment by WRMSRwasTaken** [comment]
I had the following in my config 
```
    ssl-default-bind-ciphers ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES128-GCM-SHA256
    ssl-default-bind-ciphersuites TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256
```
    
which seemed to work fine (offering both TLSv1.2 and TLSv1.3) till... I can't really tell :( But then I noticed by accident that my HAProxy doesn't offer TLS1.2 anymore (guess TLSv1.2 isn't really much needed for my clients anymore) but I still like to keep it for a while before going TLSv1.3 only.

To be honest, I am still not 100% understanding both config options with the new doc. So apparently `ssl-default-bind-ciphers` is getting completely ignored now when `ssl-default-bind-ciphersuites` has been set?

So I tried and merged both together:
```
ssl-default-bind-ciphersuites TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES128-GCM-SHA256
```
with `ssl-default-bind-ciphers` being unset which now re-enabled TLSv1.2 and TLSv1.3, but the Qualys SSL Report now shows my server offering Ciphers I don't even have set:

![image](https://github.com/haproxy/haproxy/assets/4244869/70ad0603-13b7-4153-a6a9-d013faf9e594)

I'm sorry, if I'm missing something here, but how am I supposed to disable the CBC mode now? And where's my RSA isntead of ECDSA?

It seems that OpenSSL is doing OpenSSL things, because for example on Dovecot I am having the same problem now, and it has only one config option to set the TLS ciphers `ssl_cipher_list`. I was able to restore TLSv1.2 there aswell (with the same cipher list as above), but now also offering more Ciphers than I've set. This might've been all caused when I did the switch to OpenSSL version 3.2? As said, I can't really tell because it really seems that none of my clients still depend on TLSv1.2.

€: Hmmm, setting
```
    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305
    ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
```
in HAProxy and for example in Dovecot's config aswell:
```
ssl_cipher_list = ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305
```
brings back TLSv1.2 without Qualys complaining about weak ciphers, but RSA signatures are completely missing now, only ECDSA is being offered now... looking at the docs and migration guide I can't find anything about this, maybe this is a result in their change about "providers"? I also can't spot anything in my distro's (Arch Linux) OpenSSL configure flags. Also looking at `openssl ciphers -s -v DEFAULT:@SECLEVEL=2` still lists the ciphers with RSA signatures.

**Comment by wlallemand** [comment]
@WRMSRwasTaken This really looks like you are trying to use an ECDSA key with RSA ciphers.

I made some tests with OpenSSL 3.2.1:

```
global
	ssl-default-bind-ciphers ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES128-GCM-SHA256
	ssl-default-bind-ciphersuites TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256


frontend fr1
	bind *:8443 ssl crt kikyo.pem.ecdsa
```

```
testssl -E https://kikyo:8443



Hexcode  Cipher Suite Name (OpenSSL)       KeyExch.   Encryption  Bits     Cipher Suite Name (IANA/RFC)
-----------------------------------------------------------------------------------------------------------------------------
SSLv2  
SSLv3  
TLS 1  
TLS 1.1  
TLS 1.2  
TLS 1.3  
 x1302   TLS_AES_256_GCM_SHA384            ECDH 253   AESGCM      256      TLS_AES_256_GCM_SHA384                             
 x1303   TLS_CHACHA20_POLY1305_SHA256      ECDH 253   ChaCha20    256      TLS_CHACHA20_POLY1305_SHA256                       
 x1301   TLS_AES_128_GCM_SHA256            ECDH 253   AESGCM      128      TLS_AES_128_GCM_SHA256           
```

Then changing the configuration to use an RSA private key:

```
Hexcode  Cipher Suite Name (OpenSSL)       KeyExch.   Encryption  Bits     Cipher Suite Name (IANA/RFC)
-----------------------------------------------------------------------------------------------------------------------------
SSLv2  
SSLv3  
TLS 1  
TLS 1.1  
TLS 1.2  
 xc030   ECDHE-RSA-AES256-GCM-SHA384       ECDH 253   AESGCM      256      TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384              
 xcca8   ECDHE-RSA-CHACHA20-POLY1305       ECDH 253   ChaCha20    256      TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256        
 xc02f   ECDHE-RSA-AES128-GCM-SHA256       ECDH 253   AESGCM      128      TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256              
TLS 1.3  
 x1302   TLS_AES_256_GCM_SHA384            ECDH 253   AESGCM      256      TLS_AES_256_GCM_SHA384                             
 x1303   TLS_CHACHA20_POLY1305_SHA256      ECDH 253   ChaCha20    256      TLS_CHACHA20_POLY1305_SHA256                       
 x1301   TLS_AES_128_GCM_SHA256            ECDH 253   AESGCM      128      TLS_AES_128_GCM_SHA256                      
```

Note that OpenSSL 3.2.0 is just ignoring wrong ciphersuites unlike 1.1.1 which is emitting an error.

Setting the TLS1.2 IANA ciphersuites with the `ciphersuites` keyword does not seem to behave correctly from my tests...   if you use only the ciphersuites keyword for TLSv1.2 without anything with the ciphers keyword, it looks like there is a default initialization which is bothersome, I'll try to fix this.


> `ssl-default-bind-ciphersuites TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES128-GCM-SHA256`

Regarding this line, it can't work because you used OpenSSL ciphersuites format instead of IANA ciphersuites format (TLS_ECDHE_RSA... vs ECDHE-RSA...) and OpenSSL 3.2.0 is simply ignoring this, where 1.1.1 would emit an error.

**Comment by wlallemand** [comment]
Honestly this OpenSSL behavior is confusing, the doc is unclear and from my tests using TLS1.2 IANA ciphersuites only is not possible. It's not related to haproxy at all, openssl s_server has the same problem:

Setting a TLSv1.2 ciphersuite only won't do anything:
```
$ openssl s_server -ciphersuites TLS_RSA_WITH_AES_128_CBC_SHA -accept 8443 -www -cert kikyo.pem.rsa -key kikyo.pem.rsa 
```
```
$ testssl -E https://kikyo:8443
Hexcode  Cipher Suite Name (OpenSSL)       KeyExch.   Encryption  Bits     Cipher Suite Name (IANA/RFC)
-----------------------------------------------------------------------------------------------------------------------------
SSLv2  
SSLv3  
TLS 1  
TLS 1.1  
TLS 1.2  
TLS 1.3  
```

Setting a TLSv1.2 cipher + a TLSv1.2 ciphersuite will complete the TLSv1.2 cipher list:
```
$ openssl s_server -cipher ECDHE-RSA-CHACHA20-POLY1305 -ciphersuites TLS_AES_128_GCM_SHA256:TLS_RSA_WITH_AES_128_CBC_SHA -accept 8443 -www -cert kikyo.pem.rsa -key kikyo.pem.rsa
```

```
$ testssl -E https://kikyo:8443
Hexcode  Cipher Suite Name (OpenSSL)       KeyExch.   Encryption  Bits     Cipher Suite Name (IANA/RFC)
-----------------------------------------------------------------------------------------------------------------------------
SSLv2  
SSLv3  
TLS 1  
TLS 1.1  
TLS 1.2  
 xcca8   ECDHE-RSA-CHACHA20-POLY1305       ECDH 253   ChaCha20    256      TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256        
 x2f     AES128-SHA                        RSA        AES         128      TLS_RSA_WITH_AES_128_CBC_SHA                       
TLS 1.3  
 x1301   TLS_AES_128_GCM_SHA256            ECDH 253   AESGCM      128      TLS_AES_128_GCM_SHA256           
```

So in my opinion this TLSv1.2 IANA ciphersuites is an openssl bug, it shouldn't be supported. I'll try to clarify the HAProxy documentation about this. Even in this ticket https://github.com/openssl/openssl/issues/11899 they are saying that this is not supported.
